성능 개선 및 구조, Stock Entitiy 추가 진행 중 입니다.

tux31337 opened the pull request:
### 요청사항
성능 개선 및 구조, Stock Entitiy 추가 진행 중 입니다.

### 변경사항
- Redis 개선

NotificationStatusRepository Redis 성능 최적화
Redis KEYS 명령어 제거 및 Secondary Index 패턴 적용

- 로깅 개선
1. MdcLoggingFilter
    - 모든 HTTP 요청에 고유 requestId 자동 생성 (UUID 8자리)
    - @Order(HIGHEST_PRECEDENCE)로 가장 먼저 실행
    - 요청 종료 시 MDC.clear()로 메모리 정리

 
 2. JwtAuthenticationFilter
    - JWT 인증 성공 시 userId를 MDC에 자동 추가
    - MDC.put("userId", userId)


- 구조 수정

 [Transaction]
 - repository 패키지 신규 생성
 - TransactionRepository interface 이동 (domain → repository)
 - JpaTransactionRepositoryAdapter → JpaTransactionRepository 변경
 - 이벤트 발행 Service Layer에서 발행
 [User]
 - repository 패키지 신규 생성
 - UserRepository, EmailVerificationRepository interface 이동
 - JpaUserRepositoryAdapter → JpaUserRepository 변경
 - RedisEmailVerificationRepository 이동 (infrastructure → repository)
 

- Stock 도메인 추가 진행 중
DDD 기반 Stock 도메인 구조 구현

- common/model/vo/에 공통 VO 이동 (StockCode, Money, Price, Quantity)
- Stock AggregateRoot 생성 (가격 관리 및 도메인 로직)
- ValueObjects 구현 (StockName, Sector, CurrentPrice)
- Market Enum 추가 (한국/미국 증시 지원)
- 거장 분석을 위한 섹터/시장 분류 로직 포함


--------------------------------------------------------------------------------------------------------------------
[Stock]

#### 도메인 구조
- **DDD 기반 Stock 도메인 구조 구현**
- `common/model/vo/`에 공통 VO 이동 (StockCode, Money, Price, Quantity)
- Stock AggregateRoot 생성 (가격 관리 및 도메인 로직)
- ValueObjects 구현 (StockName, Sector, CurrentPrice)
- Market Enum 추가 (한국/미국 증시 지원)
- 거래 분석을 위한 섹터/시장 분류 로직 포함

#### 실시간 주가 수신 (WebSocket)
- **KisWebSocketClient** 구현
  - KIS WebSocket 연결 및 Approval Key 발급
  - 실시간 체결가 수신 및 파싱
  - 종목 구독/해제 기능
  - 자동 재연결 기능
- **KisRealtimePriceData** 파싱 로직
  - 파이프(`|`) 및 캐럿(`^`) 구분 데이터 처리
- **KisApprovalResponse**, **KisWebSocketResponse** DTO 추가

#### 실시간 데이터 전송 (SSE)
- **SseEmitterManager** 구현
  - 종목별 SSE 연결 관리
  - 클라이언트 실시간 데이터 전송
  - 연결 상태 관리 (타임아웃 30분)
- **StockController** SSE 엔드포인트 추가
  - `GET /api/stocks/{stockCode}/price/stream`
  - WebSocket 구독 자동 연결

#### 이벤트 기반 아키텍처
- **StockPriceUpdateEvent** 도메인 이벤트 생성
- **StockPriceUpdateEventListener** 구현
  - 이벤트 수신 → SSE 전송
  - 히스토리 버퍼링 (5초마다 일괄 저장)
  - 비동기 처리 (`@Async`)

#### 가격 히스토리 저장
- **StockPriceHistory** 엔티티 생성
  - 실시간 체결가 히스토리 저장
  - 인덱스: `(stock_code, trade_time)`, `trade_time`
- **StockPriceHistoryRepository** 생성
  - 시간 범위 조회
  - 최근 N개 조회

#### 초기화 로직
- **StockDataInitializer** 구현
  - 앱 시작 시 자동 실행 (`@PostConstruct`)
  - 주요 종목 14개 자동 등록
  - 초기 가격 조회 (REST API)
  - WebSocket 구독 시작
  - Profile: `local` 환경에서만 활성화

#### 스케줄러
- **StockPriceScheduler** 구현
  - WebSocket 헬스 체크 (1분마다)
  - 장 시작 전 가격 업데이트 (주중 08:50)
  - 장 마감 후 가격 보정 (주중 15:35)
  - WebSocket 실패 시 REST API 백업

#### 설정 파일
- **AsyncConfig** 추가
  - 비동기 처리 활성화 (`@EnableAsync`)
- **CacheConfig** 추가
  - Caffeine 캐시 설정
  - 주식 가격 캐싱 (최대 500개, 1분 만료)

#### API Provider 개선
- **KisApiStockPriceProvider** 개선
  - Rate Limit 방지 로직 추가 (500ms 간격)
  - Profile 변경: `prod` → `local/dev/test`
  - 마지막 API 호출 시각 추적
- **MockStockPriceProvider** Profile 변경
  - `dev/test` → `prod` (반대)

#### Response 모델
- **StockPriceUpdateResponse** 추가
  - 실시간 가격 업데이트 응답 DTO
- **StockDetailResponse** 변경
  - 필드명 변경: `realtimePrice` → `currentPrice`, `savedPrice` → `previousPrice`
  - 필드 추가: `createdAt`
- **StockResponse** 변경
  - 필드 추가: `createdAt`

#### 의존성 추가
- **controller/build.gradle**: WebSocket 의존성 추가
- **service/stock/build.gradle**: WebSocket, Jackson, Caffeine Cache 의존성 추가

--------------------------------------------------------------------------------------------------------------------
#### 변경사항
- **Flyway** 적용
- Stock Data 오래된 경우 조회만 했는데, DB Update 까지 적용
-------------------------------------------------------------------------------------------------------------------
#### 변경사항
- TimeScaleDB 적용
- Order, Portfolio  도메인 추가
  - Order 도메인: 주문 시스템 (시장가/지정가, 부분 체결, 평균 체결가 계산)
  - Portfolio 도메인: 잔고/보유 종목 관리 (매수/매도, 평균 단가 계산)
  - Value Objects 개선: Money.ZERO, Quantity.isZero(), Price 비교 메서드 추가
  - Order Sheduling : Order ShedulingConfig 추가
  - OrderMatchingEngine : 1초마다 실행형식으로 체결조건을  계속 검사 후, 조건 충족 시 주문 실행.
  ---------------------------------------------------------------------------------------------------------------
  ### 예외 처리 개선
- ErrorCode Enum 패턴 적용
- ErrorCode 인터페이스 생성 (HttpStatus, LogLevel 포함)
- 도메인별 ErrorCode Enum 구현
  - OrderErrorCode
  - PortfolioErrorCode
  - StockErrorCode
  - EmailVerificationErrorCode
- BusinessException 리팩토링
  - ErrorCode 기반 HttpStatus / LogLevel 자동 추출
- GlobalExceptionHandler 단순화
  - if-else 제거
  - 단일 핸들러로 통합

### 공통 엔티티 개선
- BaseEntity 구현
  - @MappedSuperclass 기반 공통 엔티티 추상 클래스
- JPA Auditing 적용
  - @CreatedDate
  - @LastModifiedDate
- JpaConfig 추가
  - @EnableJpaAuditing
- BaseEntity 적용
  - common/model Entity 확장을 통해 전체 AggregateRoot에 자동 적용
- 적용 엔티티
  - Portfolio
  - Order
  - User
  - Stock
  - Transaction
  - Holding
- 중복 필드 및 로직 제거
  - createdAt, updatedAt
  - 수동 설정 로직 제거

### 낙관적 락 적용
- Order 엔티티
  - @Version 필드 추가 (동시성 제어)

### Flyway 마이그레이션 수정
- V1__create_tables.sql 수정
  - 누락된 테이블 추가
    - portfolios
    - holdings
    - orders
- holdings 테이블 컬럼명 변경
  - first_purchased_at → created_at

### 설정 수정
- PrimaryDataSourceConfig
  - hibernate.hbm2ddl.auto
    - none → create-drop

### 버그 수정
- OrderController
  - 컴파일 에러 수정
    - getTotalFilledAmount() → getTotalAmount()

### 페이징 처리 적용
- StockController
  - /search 엔드포인트 Page<> 반환으로 변경
  - /market/{market} 엔드포인트 Page<> 반환으로 변경
  - 기본 페이지 크기: 20
- StockRepository / JpaStockRepository
  - 페이징 메서드 추가
    - findByMarket
    - searchByNameContaining
- StockApplicationService
  - 페이징 처리 로직 추가

tux31337 committed:
Fix : Executor Bean 설정

tux31337 committed:
Fix : NotificationStatusRepository Redis 성능 최적화
Redis KEYS 명령어 제거 및 Secondary Index 패턴 적용

tux31337 committed:
Refactor : 로깅 개선 및 JWT 토큰 개선
[로깅 개선]
1. MdcLoggingFilter
   - 모든 HTTP 요청에 고유 requestId 자동 생성 (UUID 8자리)
   - @Order(HIGHEST_PRECEDENCE)로 가장 먼저 실행
   - 요청 종료 시 MDC.clear()로 메모리 정리

2. JwtAuthenticationFilter
   - JWT 인증 성공 시 userId를 MDC에 자동 추가
   - MDC.put("userId", userId)

3. MdcTaskDecorator
   - 비동기 작업(@Async)에 MDC 컨텍스트 전파
   - ThreadLocal 기반 MDC를 새 스레드에 복사
   - SecurityContext도 함께 전파

4. logback-spring.xml
   - 로그 패턴에 [%X{requestId}] [%X{userId}] 추가
   - 모든 로그에 자동으로 추적 정보 출력

5. 비동기 Config 적용
   - NotificationAsyncConfig
   - TransactionAsyncConfig
   - executor.setTaskDecorator(new MdcTaskDecorator())

[JWT 최적화]
 - email 이 아닌 userId 로 바꿈

tux31337 committed:
Refactor : Repository 패키지 추가 및 불필요한 Adapter 제거, Transaction Entity 수정
[Transaction]
- repository 패키지 신규 생성
- TransactionRepository interface 이동 (domain → repository)
- JpaTransactionRepositoryAdapter → JpaTransactionRepository 변경
- 이벤트 발행 Service Layer에서 발행
[User]
- repository 패키지 신규 생성
- UserRepository, EmailVerificationRepository interface 이동
- JpaUserRepositoryAdapter → JpaUserRepository 변경
- RedisEmailVerificationRepository 이동 (infrastructure → repository)

tux31337 committed:
Feat(stock): DDD 기반 Stock 도메인 구조 구현

- common/model/vo/에 공통 VO 이동 (StockCode, Money, Price, Quantity)
- Stock AggregateRoot 생성 (가격 관리 및 도메인 로직)
- ValueObjects 구현 (StockName, Sector, CurrentPrice)
- Market Enum 추가 (한국/미국 증시 지원)
- 거장 분석을 위한 섹터/시장 분류 로직 포함

tux31337 committed:
Feat : Stock 데이터 제공
 - StockPriceProvider 인터페이스 통해 도메인이 인프라에 의존하지 않게
 - MockStockProvider 생성 - 실제 API 연동하지 않고 개발용으로 사용
 - kis 연동 진행 중

tux31337 committed:
Feat : Stock Domain 작업 중
 - 한국투자증권 API 연동 중
 - 필요한 StockSearch, Detail, Stock Response DTO 생성
 - Service 계층 작업 중
🔍 조회 (Query)
searchStocks(keyword) - 종목명 검색 (자동완성)
getStockDetail(stockCode) - 종목 상세 조회 (실시간 가격 포함)
getStocksByMarket(market) - 시장별 종목 조회
getStocksBySector(sector) - 섹터별 종목 조회
getStocks(stockCodes) - 여러 종목 일괄 조회
getAllStocks() - 전체 종목 조회
existsStock(stockCode) - 종목 존재 여부 확인
countByMarket(market) - 시장별 종목 수

등록/수정 (Command)

registerStock(...) - 종목 등록 (관리자)
updateStockPrice(stockCode) - 특정 종목 가격 업데이트
updateAllStockPrices() - 전체 종목 가격 일괄 업데이트 (배치)
deleteStock(stockCode) - 종목 삭제 (관리자)

tux31337 committed:
Feat : Stock 관련 작업 중
 - Controller 작성
 - KIS API 오류 작업 진행 중

tux31337 committed:
Feat(stock): add WebSocket health check and improve initialization flow

tux31337 committed:
Feat: StockAsyncConfig 분리

tux31337 committed:
Feat : 환경별 설정 수정 및 Redis 추가, Redis 최우선 조회로 수정

tux31337 committed:
Feat: flyway 적용 및 데이터가 오래된 경우 데이터를 가지고오고 데이터베이스 최신화 로직 적용

tux31337 committed:
Feat : Stock Exception Global Exception 으로 변경

tux31337 committed:
Feat : Stock Exception

tux31337 committed:
Feat : TimeScale DB 적용

tux31337 committed:
Feat: Order 및 Portfolio 도메인 구현 (실제 거래소 시스템 구축)
구현 내용
- Order 도메인: 주문 시스템 (시장가/지정가, 부분 체결, 평균 체결가 계산)
- Portfolio 도메인: 잔고/보유 종목 관리 (매수/매도, 평균 단가 계산)
- Value Objects 개선: Money.ZERO, Quantity.isZero(), Price 비교 메서드 추가

tux31337 committed:
Feat: Order 및 Portfolio 도메인 구현 (실제 거래소 시스템 구축)
구현 내용
- Order Sheduling : Order ShedulingConfig 추가
- OrderMatchingEngine : 1초마다 실행형식으로 체결조건을  계속 검사 후, 조건 충족 시 주문 실행.

hoho-green commented on controller/src/main/java/com/truvis/controller/stock/StockController.java:
```diff
@@ -1,20 +1,132 @@
+     * 종목 검색 (자동완성)
+     * GET /api/stocks/search?keyword=삼성
+     */
+    @GetMapping("/search")
```
p3. 페이징 처리 되어야함 이게 결과 값이 많으면 큰일날수있음

hoho-green commented on service/common/src/main/java/com/truvis/common/config/GlobalExceptionHandler.java:
```diff
@@ -55,6 +56,43 @@ public ResponseEntity<ErrorResponse> handleBusinessException(BusinessException e
+        HttpStatus status;
+        
+        // 에러 코드에 따라 HTTP 상태 코드 결정
+        if ("STOCK_001".equals(errorCode)) {
```
p3. if ~ else 를 줄이는 방법을 고민해주세요 ~! 패턴 사용하면 될듯 팩토리 패턴이나

hoho-green commented on service/portfolio/src/main/java/com/truvis/portfolio/domain/Holding.java:
```diff
@@ -0,0 +1,228 @@
+@Table(name = "holdings")
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class Holding {
```
p3. base entity  구현해서 사용

tux31337 committed:
Feat : User Column 변경 및 OrderMatchingEngine 실제 데이터로 변경

tux31337 committed:
Feat : 프로젝트 구조 변경 및 타임스케일DB설정

tux31337 committed:
Feat: Portfolio 서비스 완성 및 Order 연동, Security 적용
- PortfolioApplicationService: 생성/조회/분석, 입출금, 매수/매도 검증
- Order-Portfolio 연동: 주문 시 잔고/보유량 실시간 검증
- REST API: PortfolioController, OrderController 추가
- Security 연동: SecurityUtils로 인증된 userId 추출
- 공통 예외: PortfolioException, OrderException, GlobalExceptionHandler

tux31337 committed:
Refactor: ErrorCode Enum 패턴 적용으로 GlobalExceptionHandler 단순화
- ErrorCode 인터페이스 추가 (HttpStatus, LogLevel 포함)
- 도메인별 ErrorCode Enum 생성
  - OrderErrorCode, PortfolioErrorCode, StockErrorCode, EmailVerificationErrorCode
- BusinessException이 ErrorCode 기반으로 HttpStatus 자동 결정
- GlobalExceptionHandler에서 if-else 제거, 통합 핸들러로 단순화

tux31337 committed:
refactor: BaseEntity 구현 및 오류 수정
1. BaseEntity 구현
   - createdAt, updatedAt 자동 관리 추상 클래스
   - JpaConfig에 @EnableJpaAuditing 설정
   - Holding 엔티티에 적용 (firstPurchasedAt → createdAt), Order, Portfolio, Stock 적용
3. DB 마이그레이션 정비
   - V1에 portfolios, holdings, orders 테이블 추가
2. 버그 수정
   - OrderController: getTotalFilledAmount() → getTotalAmount()

tux31337 committed:
Feat(stock): 종목 검색/시장별 조회 API에 페이징 적용

- StockController: /search, /market/{market} 엔드포인트에 Page 반환
- StockRepository: 페이징 메서드 추가 (findByMarket, searchByNameContaining)
- StockApplicationService: 페이징 처리 로직 추가

API 변경:
- GET /api/stocks/search?keyword=삼성 → Page<StockSearchResponse>
- GET /api/stocks/market/{market} → Page<StockResponse>
- 기본 페이지 크기: 20, ?page=0&size=N 으로 조절 가능

hoho-green commented on controller/src/main/java/com/truvis/controller/order/OrderController.java:
```diff
@@ -0,0 +1,292 @@
+                userId, request.getStockCode(), request.getOrderType(), request.getQuantity());
+
+        Order order;
+        if (request.getOrderType() == OrderType.MARKET) {
```
p1. 이부분 리팩토링 한번 해보세요 if else 가 너무 많이 있음
